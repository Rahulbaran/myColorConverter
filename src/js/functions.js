/**
 * Convert HEX color value to RGB
 * returns an array of Red, Green & Blue
 *
 * @param  String  Hex value
 * @return Array  The r,g,b values
 */
export const hexToRgb = hex => {
  if (hex.length === 4) {
    return hex
      .slice(1)
      .split("")
      .map(e => parseInt(e, 16));
  } else if (hex.length === 7) {
    return hex
      .slice(1)
      .match(/[\da-f]{2}/gi)
      .map(e => parseInt(e, 16));
  }
};

/**
 * Convert RGB color value to Hex
 * returns a string of hex color value
 *
 * @param  Array   [r,g,b]  red, green & blue values
 * @return String  Hex Value
 */
export const rgbToHex = rgb => {
  const rgbValues = rgb
    .slice(4)
    .slice(0, -1)
    .trim()
    .split(/,\s{0,}/)
    .filter(e => e <= 255);

  return rgbValues.length === 3
    ? "#" +
        rgbValues
          .map(e =>
            (+e).toString(16).length < 2
              ? "0" + (+e).toString(16)
              : (+e).toString(16)
          )
          .join("")
    : null;
};

/**
 * Converts RGB color value into HSL.
 * returns an array containing h,s & l
 *
 * @param  Array   [r,g,b]  red, green & blue values
 * @return Array   The hue, stauration & lightness values
 */
export const rgbToHsl = (...rgbArr) => {
  let h, s;

  // Calculate fraction of r,g,b
  const rgb = rgbArr.map(val => val / 255);
  const [r, g, b] = rgb;

  // Calculate lightness
  const [min, max] = [Math.min(...rgb), Math.max(...rgb)];
  const l = (min + max) / 2;

  // Calculate hue & saturation
  const delta = max - min;
  if (delta === 0) h = s = 0;
  else {
    s = l > 0.5 ? delta / (2 - l * 2) : delta / (l * 2);

    switch (max) {
      case r:
        h = (g - b) / delta + (g < b ? 6 : 0);
        break;
      case g:
        h = (b - r) / delta + 2;
        break;
      case b:
        h = (r - g) / delta + 4;
        break;
    }
    h = Math.round(h * 60);
  }

  return [h, +(s * 100).toFixed(0), +(l * 100).toFixed(0)];
};
