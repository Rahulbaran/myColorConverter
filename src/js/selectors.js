export const selectors = {
  $hexToRgbBtn: document.querySelector(".hextorgb__btn"),
  $rgbToHexBtn: document.querySelector(".rgbtohex__btn"),
  $hexToRgbCard: document.querySelector(".hextorgb__card"),
  $rgbToHexCard: document.querySelector(".rgbtohex__card"),
  $hexInputField: document.querySelector(".hex__input--field"),
  $rgbInputField: document.querySelector(".rgb__input--field"),
  $firstColorWindow: document.querySelector(".card:first-of-type .card__color--window"),
  $lastColorWindow: document.querySelector(".card:last-of-type .card__color--window"),
  $copyBtns: document.querySelectorAll(".btn .bxs-copy"),
  $hexColorLabel: document.querySelector(".hex__value span"),
  $rgbColorLabel: document.querySelector(".rgb__value span"),
  $firstHslColorLabel: document.querySelector(".hextorgb__card .hsl__value span"),
  $lastHslColorLabel: document.querySelector(".rgbtohex__card .hsl__value span")
};
