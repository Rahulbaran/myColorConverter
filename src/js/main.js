import "../scss/main.scss";
import { selectors as sel } from "./selectors";
import { rgbToHex, hexToRgb, rgbToHsl } from "./functions";

const [hexRegex, rgbRegex] = [
  /^#([\da-f]{3}|[\da-f]{6})$/i,
  /^(rgb)\(\s{0,}(\d{1,3}\s{0,},\s{0,}){2}\d{1,3}\s{0,}\)$/i
];

/*----------- Event handler for card switch buttons -----------*/
sel.$hexToRgbBtn.addEventListener("click", function () {
  // Toggle class in buttons
  if (!this.classList.contains("btn__active"))
    this.classList.add("btn__active");
  sel.$rgbToHexBtn.classList.remove("btn__active");

  // Toggle class in cards
  if (!sel.$hexToRgbCard.matches(".card-visible"))
    sel.$hexToRgbCard.classList.add("card-visible");
  sel.$rgbToHexCard.classList.remove("card-visible");
});

sel.$rgbToHexBtn.addEventListener("click", function () {
  // Toggle class in buttons
  this.classList.toggle("btn__active");
  sel.$hexToRgbBtn.classList.toggle("btn__active");

  // Toggle class in cards
  sel.$rgbToHexCard.classList.add("card-visible");
  sel.$hexToRgbCard.classList.remove("card-visible");
});

/*--------------- Event handler for copy buttons --------------*/
try {
  sel.$copyBtns.forEach(cpBtn => {
    cpBtn.addEventListener("click", function () {
      navigator.clipboard
        .writeText(this.parentElement.previousElementSibling.textContent)
        .then(() => alert("color value has been copied"));
    });
  });
} catch (error) {
  sel.$copyBtns.forEach(cpBtn => (cpBtn.style.display = "none"));
}

/*-------------- Event Handler for hex input field --------------*/
sel.$hexInputField?.addEventListener("input", function () {
  if (hexRegex.test(this.value.trim())) {
    // Calculate rgb & hsl values
    const [r, g, b] = hexToRgb(this.value.trim());
    const [h, s, l] = rgbToHsl(r, g, b);

    // Display rgb & hex values in UI
    const rgbColorValue = `rgb(${r}, ${g}, ${b})`;
    sel.$rgbColorLabel.textContent = rgbColorValue;
    sel.$firstHslColorLabel.textContent = `hsl(${h}, ${s}%, ${l}%)`;

    // Display calculated color in color window
    sel.$firstColorWindow.style.backgroundColor = rgbColorValue;
  }
});

/*--------------- Event Handler for rgb input field ---------------*/
sel.$rgbInputField?.addEventListener("input", function () {
  if (rgbRegex.test(this.value.trim())) {
    // Calculate hex & hsl values
    const hexColor = rgbToHex(this.value.trim());
    if (hexColor) {
      const [h, s, l] = rgbToHsl(...hexToRgb(hexColor));

      // Display hex & hsl values in UI
      sel.$hexColorLabel.textContent = hexColor;
      sel.$lastHslColorLabel.textContent = `hsl(${h}, ${s}%, ${l}%)`;

      // Display calcualted color value in color window
      sel.$lastColorWindow.style.backgroundColor = hexColor;
    }
  }
});
