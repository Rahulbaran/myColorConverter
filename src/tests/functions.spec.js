import { describe, expect, test } from "vitest";
import { hexToRgb, rgbToHex, rgbToHsl } from "../js/functions";

// * Tests for hexToRgb() function
describe("hexToRgb", () => {
  test("returns an array of r,g,b values", () => {
    const rgbArray = hexToRgb("#123");

    expect(rgbArray).toHaveLength(3);
    expect(rgbArray).toEqual([1, 2, 3]);
  });

  test("returns an array of r,g,b values", () => {
    expect(hexToRgb("#123fad")).toEqual([18, 63, 173]);
  });
});

// * Tests for rgbToHex() function
describe("rgbToHex", () => {
  test("returns a string containing hex color value", () => {
    const hexValue = rgbToHex("rgb(1, 46, 5)");

    expect(hexValue).toHaveLength(7);
    expect(hexValue).toBe("#012e05");
  });

  test("returns null as value", () => {
    expect(rgbToHex("rgb(433, 345, 53)")).toBeNull();
  });

  test("generate snapshot", () => {
    const bool = rgbToHex("rgb(1, 46, 5)") === "#012e05";
    expect(bool).toMatchSnapshot();
  });
});

// * Tests for rgbToHsl() function
describe("rgbToHsl", () => {
  test("returns an array containing hue, stauration & lightness values", () => {
    const hslValue = rgbToHsl(17, 134, 156);

    expect(hslValue).toHaveLength(3);
    expect(hslValue).toEqual([189, 80, 34]);
  });
});
