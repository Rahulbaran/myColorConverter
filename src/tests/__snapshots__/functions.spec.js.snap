// Vitest Snapshot v1

exports[`rgbToHex > generate snapshot 1`] = `true`;
